### ISSUE TITLE
Singleton pattern type incompatible with singleton-typed scrutinee

### ISSUE THREAD

**Issue by milessabin** [scala/bug#10569]
A type pattern `Singleton` is treated as incompatible with a singleton-typed scrutinee,

```scala
Tuple1[Foo.type](Foo) match {
  case Tuple1(foo: Singleton) => foo
}
```
```
error: pattern type is incompatible with expected type;
 found   : Singleton
 required: Test.Foo.type
    case Tuple1(foo: Singleton) => foo
       ^
one error found
```
Fixed in https://github.com/scala/scala/pull/6150.

This turned out to be the cause of https://github.com/typelevel/scala/issues/152 reported against the SIP-23 literal types extension, but as can be seen the example here it's an issue for singleton types generally.

**Comment by adriaanm** [comment]
Fixed in scala/scala#6150.

**Comment by dwijnand** [comment]
Fixed in https://github.com/scala/scala/pull/6150.
